source/driver_run.cpp

```diff
@@ -2,86 +2,134 @@
 #include "module_cell/module_neighbor/sltk_atom_arrange.h"
 #include "module_hamilt_pw/hamilt_pwdft/global.h"
 #include "module_io/input.h"
+#include "module_io/para_json.h"
 #include "module_io/print_info.h"
 #include "module_io/winput.h"
 #include "module_md/run_md.h"
-#include "module_io/para_json.h"
 
+#ifdef __LCAO
+#include "module_beyonddft/esolver_lrtd_lcao.h"
+#endif
+extern "C"
+{
+#include "module_base/blacs_connector.h"
+}
 /**
  * @brief This is the driver function which defines the workflow of ABACUS calculations.
  * It relies on the class Esolver, which is a class that organizes workflows of single point calculations.
- * 
+ *
  * For calculations involving change of configuration (lattice parameter & ionic motion),
  * this driver calls Esolver::Run and the configuration-changing subroutine in a alternating manner.
- * 
+ *
  * Information is passed between the two subroutines by class UnitCell
- * 
- * Esolver::Run takes in a configuration and provides force and stress, 
+ *
+ * Esolver::Run takes in a configuration and provides force and stress,
  * the configuration-changing subroutine takes force and stress and updates the configuration
  */
 void Driver::driver_run(void)
 {
     ModuleBase::TITLE("Driver", "driver_line");
     ModuleBase::timer::tick("Driver", "driver_line");
 
-    //! 1: initialize the ESolver 
-    ModuleESolver::ESolver *p_esolver = nullptr;
+    //! 1: initialize the ESolver
+    ModuleESolver::ESolver* p_esolver = nullptr;
     ModuleESolver::init_esolver(p_esolver);
 
     //! 2: setup cell and atom information
 
     // this warning should not be here, mohan 2024-05-22
 #ifndef __LCAO
-    if(GlobalV::BASIS_TYPE == "lcao_in_pw" || GlobalV::BASIS_TYPE == "lcao")
+    if (GlobalV::BASIS_TYPE == "lcao_in_pw" || GlobalV::BASIS_TYPE == "lcao")
     {
-        ModuleBase::WARNING_QUIT("driver","to use LCAO basis, compile with __LCAO");
+        ModuleBase::WARNING_QUIT("driver", "to use LCAO basis, compile with __LCAO");
     }
 #endif
 
     // the life of ucell should begin here, mohan 2024-05-12
     // delete ucell as a GlobalC in near future
     GlobalC::ucell.setup_cell(GlobalV::stru_file, GlobalV::ofs_running);
 
-    //! 3: initialize Esolver and fill json-structure 
-    p_esolver->before_all_runners(INPUT, GlobalC::ucell);
+    //! 3: initialize Esolver and fill json-structure
+#ifdef __LCAO
+    if (GlobalV::ESOLVER_TYPE == "lr")
+        // use constructor rather than Init function to initialize reference (instead of pointers) to ucell
+        if (GlobalV::GAMMA_ONLY_LOCAL)
+            p_esolver = new ModuleESolver::ESolver_LRTD<double, double>(INPUT, GlobalC::ucell);
```

Comment:
Please move this part to init_esolver function. This cpp file should not include anything about LCAO.  This file should not have "#ifndef __LCAO" or "ifdef __LCAO".

Reply:
driver should not depend on LCAO, it not depend on esolver

Reply:
I don't think #ifdef-#endif means dependence, because the dependent header `esolver_lrtd_lcao` is also in #ifdef-#endif.
Also, there's another `__LCAO` in this file:
https://github.com/deepmodeling/abacus-develop/blob/develop/source/driver_run.cpp#L34-L39

Reply:
My `ESolver_LRTD` depends on an **initialized** `ucell`. 
I will move `init_esolver` after `ucell.setup_cell` and make it depend on `ucell`, which is a reasonable order and dependence. If you agree, then this part can be moved into `init_esolver`.

Reply:
Of course, every esolver depends on ucell